### PR TITLE
Fix SequentialLR wrong learning rate initialization when milestones contain 0

### LINKED ISSUE #185872: SequentialLR initializes with the wrong learning rate when milestones contain 0 (due to off by one error)
### 🐛 Describe the bug

# Issue

There's an issue with `SequentialLR(..., milestones=[0, ...])`. \
We use this sometimes to test with no warmup.


- In __init__
    - It [initializes the learning rate with the **FIRST scheduler**, regardless of what the milestones specify](https://github.com/pytorch/pytorch/blob/v2.12.0/torch/optim/lr_scheduler.py#L1168-L1170).
- Then in the step function:
    - `last_epoch` is immediately [incremented to 1](https://github.com/pytorch/pytorch/blob/v2.12.0/torch/optim/lr_scheduler.py#L1187)
    - It then [**fetches the CORRECT scheduler** by running a bisect_right to get the correct idx](https://github.com/pytorch/pytorch/blob/v2.12.0/torch/optim/lr_scheduler.py#L1188-L1189). In this case, `idx` is `1`, so it gets the second scheduler.
    - It should ideally update the learning rate immediately by entering [`if idx > 0 and self._milestones[idx - 1] == self.last_epoch:`](https://github.com/pytorch/pytorch/blob/v2.12.0/torch/optim/lr_scheduler.py#L1190)
    - **[BUG HERE]** But in this case, right after init, `self._milestones[idx - 1]` is `0` and `self.last_epoch` is `1`, so the equality fails, and it goes straight to the else block, and runs [`scheduler.step`](https://github.com/pytorch/pytorch/blob/v2.12.0/torch/optim/lr_scheduler.py#L1193).
    - This runs the correct scheduler, but with the initial learning rate set by the first scheduler.

# Reproducing

## Script
```python
import torch
import torch.optim as optim
from torch import nn
from torch.optim.lr_scheduler import LinearLR, SequentialLR

model = nn.Linear(10, 10)
optimizer = optim.Adam(model.parameters(), lr=1.0)

warmup = LinearLR(optimizer, start_factor=0.1, end_factor=0.5, total_iters=10)
decay = LinearLR(optimizer, start_factor=0.5, end_factor=0.1, total_iters=10)
scheduler = SequentialLR(optimizer, schedulers=[warmup, decay], milestones=[0])

print("Initial LR:", scheduler.get_last_lr()[0])
for _ in range(10):
    scheduler.step()
    print("Step LR:", scheduler.get_last_lr()[0])
```

## Result

- The learning rates should be between 0.5 and 0.1, but instead are between 0.05 and 0.01

```
Initial LR: 0.1
Step LR: 0.05
Step LR: 0.046000000000000006
Step LR: 0.042
Step LR: 0.038000000000000006
Step LR: 0.034
Step LR: 0.030000000000000002
Step LR: 0.026000000000000002
Step LR: 0.022000000000000002
Step LR: 0.018000000000000002
Step LR: 0.014

```



### Versions

    PyTorch version: 2.12.0+cu130                                                                                                                                                          
    Is debug build: False                                                                                                                                                                  
    CUDA used to build PyTorch: 13.0                                                                                                                                                       
    ROCM used to build PyTorch: N/A                                                                                                                                                        
                                                                                                                                                                                           
    OS: Debian GNU/Linux 12 (bookworm) (x86_64)                                                                                                                                            
    GCC version: (Debian 12.2.0-14+deb12u1) 12.2.0                                                                                                                                         
    Clang version: 14.0.6                                                                                                                                                                  
    CMake version: Could not collect                                                                                                                                                       
    Libc version: glibc-2.36                                                                                                                                                               
                                                                                                                                                                                           
    Python version: 3.11.2 (main, Apr  8 2026, 01:58:00) [GCC 12.2.0] (64-bit runtime)                                                                                                     
    Python platform: Linux-6.1.0-44-cloud-amd64-x86_64-with-glibc2.36                                                                                                                      
    Is CUDA available: False                                                                                                                                                               
    CUDA runtime version: 12.4.131                                                                                                                                                         
    CUDA_MODULE_LOADING set to: N/A                                                                                                                                                        
    GPU models and configuration: Could not collect                                                                                                                                        
    Nvidia driver version: Could not collect                                                                                                                                               
    cuDNN version: Could not collect                                                                                                                                                       
    Is XPU available: False                                                                                                                                                                
    HIP runtime version: N/A                                                                                                                                                               
    MIOpen runtime version: N/A                                                                                                                                                            
    Is XNNPACK available: True                                                                                                                                                             
    Caching allocator config: N/A                                                                                                                                                          
                                                                                                                                                                                           
    CPU:                                                                                                                                                                                   
    Architecture:                            x86_64                                                                                                                                        
    CPU op-mode(s):                          32-bit, 64-bit                                                                                                                                
    Address sizes:                           46 bits physical, 48 bits virtual                                                                                                             
    Byte Order:                              Little Endian                                                                                                                                 
    CPU(s):                                  2                                                                                                                                             
    On-line CPU(s) list:                     0,1                                                                                                                                           
    Vendor ID:                               GenuineIntel                                                                                                                                  
    Model name:                              Intel(R) Xeon(R) CPU @ 2.80GHz                                                                                                                
    CPU family:                              6
    Model:                                   85
    Thread(s) per core:                      1
    Core(s) per socket:                      2
    Socket(s):                               1
    Stepping:                                7
    BogoMIPS:                                5600.32
    Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm          
  constant_tsc rep_good nopl xtopology nonstop_tsc cpuid tsc_known_freq pni pclmulqdq ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand hypervisor lahf_lm   
  abm 3dnowprefetch invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm avx512f avx512dq rdseed adx smap clflushopt clwb        
  avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat avx512_vnni md_clear arch_capabilities
    Hypervisor vendor:                       KVM
    Virtualization type:                     full
    L1d cache:                               64 KiB (2 instances)
    L1i cache:                               64 KiB (2 instances)
    L2 cache:                                2 MiB (2 instances)
    L3 cache:                                33 MiB (1 instance)
    NUMA node(s):                            1
    NUMA node0 CPU(s):                       0,1
  
    Versions of relevant libraries:
    [pip3] nvidia-cublas==13.1.1.3
    [pip3] nvidia-cuda-cupti==13.0.85
    [pip3] nvidia-cuda-nvrtc==13.0.88
    [pip3] nvidia-cuda-runtime==13.0.96
    [pip3] nvidia-cudnn-cu13==9.20.0.48
    [pip3] nvidia-cufft==12.0.0.61
    [pip3] nvidia-curand==10.4.0.35
    [pip3] nvidia-cusolver==12.0.4.66
    [pip3] nvidia-cusparse==12.6.3.3
    [pip3] nvidia-cusparselt-cu13==0.8.1
    [pip3] nvidia-nccl-cu13==2.29.7
    [pip3] nvidia-nvjitlink==13.0.88
    [pip3] nvidia-nvtx==13.0.85
    [pip3] torch==2.12.0
    [pip3] triton==3.7.0
    [conda] Could not collect


cc @vincentqb @jbschlosser @albanD @janeyx99 @crcrpar

### PR DESCRIPTION
**Pull request by aahladc**
Fixes https://github.com/pytorch/pytorch/issues/185872
  
### Cause of the Bug:
**Step Offset:** In `SequentialLR.step()`, `last_epoch` is immediately incremented to `1` on the first step. When milestones contain `0`, `self._milestones[idx - 1] == self.last_epoch` compares `0 == 1`, evaluating to `False`. This bypasses the initialization block `_update_lr(0)` and directly executes `scheduler.step()`, causing the second scheduler to skip its initial step 0.

### Fix:
1. Use `idx = bisect_right(self._milestones, 0)` in `__init__` to dynamically identify the correct active scheduler at step 0.
2. Run `_initial_step()` only on the correct starting `self._schedulers[idx]` instead of `self._schedulers[0]`.
3. Correct `self._last_lr` to track the initial learning rate from the starting `schedulers[idx]`.
    
### Testing:
Added `test_sequentiallr_skips_preceding_schedulers_with_zero_milestone` to `test_lrscheduler.py` verifying that multiple skipped schedulers under `milestones=[0, 0, 1]` successfully start at the uncorrupted target rate.

### CODE REVIEW

**Comment by janeyx99** [comment]
@claude review this PR with the pr review skill!

**Comment by aahladc** [comment]
Updated the commit to satisfy the concerns laid out by claude bot.

**Comment by janeyx99** [comment]
@aahladc I haven't looked at this code for a while and you are more familiar--what do you think about Claude's comment about silently slicing vs just not letting people pass in a milestone of 0?

**Comment by aahladc** [comment]
@janeyx99 

- People should be able to pass a milestone of 0 imo. It's a valid input as far the API goes. Since step 0 is a valid step (unlike, say, step -10, which semantically makes no sense), milestones at step 0 should be allowed!
- I amended the commit (and force pushed it) to have a slightly different solution that bypasses that issue entirely, and should address the other comments claude bot left earlier. Please feel free to run claude bot's review again, since the old one is invalid.

**Comment by janeyx99** [comment]
@claude review again!

**Review by janeyx99 — APPROVED** [review]
thank you, i'm happy with this

**Review by mlazos — APPROVED** [review]
